File: typescript/src/render.ts
import { HEADINGS } from "./selection.js";
import { Node, SPACE, elements, postorder, textContent } from "./tree.js";
import { destination, safeUrl } from "./urls.js";

const EMPHASIS = new Map([
  ["em", "*"],
  ["i", "*"],
  ["strong", "**"],
  ["b", "**"],
  ["s", "~~"],
  ["del", "~~"],
  ["strike", "~~"],
]);
const LITERAL_TAGS = new Set(["pre", "code", "kbd", "samp"]);
const BLOCKS = new Set(
  "p div article main section header footer figure figcaption address details summary dl".split(
    " ",
  ),
);
const BLOCK_NEIGHBORS = new Set([
  ...BLOCKS,
  ...HEADINGS,
  "ul",
  "ol",
  "pre",
  "table",
  "blockquote",
  "hr",
  "li",
]);

function escape(value: string): string {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replace(/([\\`*_~[\]])/g, "\\$1")
    .replace(
      /(^|\n)(\s*)(#{1,6}(?=\s)|[-+](?=\s)|\d+[.)](?=\s)|[=~-]{3,}(?=\s|$))/g,
      (_match: string, prefix: string, space: string, marker: string) =>
        prefix +
        space +
        (/^\d/.test(marker) ? `${marker.slice(0, -1)}\\${marker.slice(-1)}` : `\\${marker}`),
    );
}

function longestBackticks(value: string): number {
  let longest = 0;
  for (const match of value.matchAll(/`+/g)) {
    longest = Math.max(longest, match[0].length);
  }
  return longest;
}

function inlineCode(node: Node): string {
  const value = textContent(node, false).replace(/\r\n|\r|\n/g, " ");
  if (!value) {
    return "";
  }
  const delimiter = "`".repeat(longestBackticks(value) + 1);
  const padding =
    value.startsWith("`") ||
    value.endsWith("`") ||
    (value.startsWith(" ") && value.endsWith(" ") && value.trim())
      ? " "
      : "";
  return delimiter + padding + value + padding + delimiter;
}

function codeLanguage(node: Node): string {
  const candidates = [node, ...elements(node)];
  if (node.parent) {
    candidates.push(node.parent);
  }
  const valid = /^[\p{L}\p{N}_.+#-]+$/u;
  for (const candidate of candidates) {
    for (const token of (candidate.attrs.get("class") ?? "").split(SPACE)) {
      if (/^(?:language|lang|highlight)-/.test(token)) {
        const value = token.slice(token.indexOf("-") + 1);
        if (valid.test(value)) {
          return value;
        }
      }
    }
    const value = candidate.attrs.get("data-language") ?? "";
    if (valid.test(value)) {
      return value;
    }
  }
  return "";
}

function fencedCode(node: Node): string {
  const value = textContent(node, false).replace(/\r\n|\r/g, "\n");
  const fence = "`".repeat(Math.max(3, longestBackticks(value) + 1));
  return `\n\n${fence}${codeLanguage(node)}\n${value}${value.endsWith("\n") ? "" : "\n"}${fence}\n\n`;
}

function joinParts(parts: string[]): string {
  const result: string[] = [];
  for (let part of parts) {
    if (!part) {
      continue;
    }
    const previous = result.at(-1);
    if (previous?.endsWith("\n") && part.startsWith("\n")) {
      result[result.length - 1] = previous.replace(/\n+$/, "");
      part = "\n\n" + part.replace(/^\n+/, "");
    }
    result.push(part);
  }
  return result.join("");
}

function wrapInline(body: string, marker: string): string {
  const content = body.trim();
  if (!content) {
    return body;
  }
  const leading = /^\s/u.test(body) ? " " : "";
  const trailing = /\s$/u.test(body) ? " " : "";
  return leading + marker + content + marker + trailing;
}

function renderLink(node: Node, body: string, base: string | null): string {
  const href = safeUrl(node.attrs.get("href") ?? "", base);
  return href && body.trim() ? `[${body.trim()}](${destination(href)})` : body;
}

function renderImage(node: Node, base: string | null): string {
  const alt = escape((node.attrs.get("alt") ?? "").replace(SPACE, " ").trim());
  const source = safeUrl(node.attrs.get("src") ?? "", base);
  return source && node.attrs.get("src") ? `![${alt}](${destination(source)})` : alt;
}

function lines(value: string): string[] {
  if (!value) {
    return [];
  }
  // Preserve the reference's line boundaries within list and quote content.
  // oxlint-disable-next-line no-control-regex
  return value.split(/\r\n|[\n\r\v\f\x1c-\x1e\x85\u2028\u2029]/u);
}

function renderList(node: Node, rendered: Map<Node, string>): string {
  const start = (node.attrs.get("start") ?? "1").trim();
  let number = /^[+-]?\d+$/.test(start) ? BigInt(start) : 1n;
  const items: string[] = [];
  for (const child of elements(node)) {
    if (child.tag !== "li") {
      continue;
    }
    if (node.tag === "ol") {
      const value = child.attrs.get("value") ?? "";
      if (/^-?\d{1,9}$/.test(value)) {
        number = BigInt(value);
      }
    }
    const marker = node.tag === "ol" ? `${number}. ` : "- ";
    const [first, ...rest] = lines(rendered.get(child)!.trim());
    if (first !== undefined) {
      const continuation = rest.map((line) => (line ? " ".repeat(marker.length) + line : ""));
      items.push([marker + first, ...continuation].join("\n"));
    }
    number++;
  }
  return items.length ? "\n" + items.join("\n") + "\n" : "";
}

function tableRows(node: Node): Node[] {
  const rows: Node[] = [];
  const stack = elements(node).reverse();
  while (stack.length) {
    const child = stack.pop()!;
    if (child.tag === "tr") {
      rows.push(child);
    } else if (["thead", "tbody", "tfoot"].includes(child.tag)) {
      stack.push(...elements(child).reverse());
    }
  }
  return rows;
}

function simpleTable(rows: Node[][]): boolean {
  if (!rows[0]?.length || rows.some((row) => row.length !== rows[0]!.length)) {
    return false;
  }
  for (const row of rows) {
    for (const cell of row) {
      if (
        (cell.attrs.get("colspan") ?? "1") !== "1" ||
        (cell.attrs.get("rowspan") ?? "1") !== "1" ||
        elements(cell).some((child) => ["table", "pre", "ul", "ol", "p"].includes(child.tag))
      ) {
        return false;
      }
    }
  }
  return true;
}

function renderTable(node: Node, rendered: Map<Node, string>): string {
  const rows = tableRows(node).map((row) =>
    elements(row).filter((cell) => cell.tag === "th" || cell.tag === "td"),
  );
  const captions = elements(node)
    .filter((child) => child.tag === "caption")
    .map((child) => rendered.get(child)!.trim());
  if (!simpleTable(rows)) {
    const output = rows.map(
      (row, index) =>
        `${index + 1}. ` +
        row.map((cell) => rendered.get(cell)!.trim().replaceAll("\n", " ")).join(" — "),
    );
    return "\n\n" + [...captions, output.join("\n")].join("\n\n") + "\n\n";
  }
  // simpleTable establishes a nonempty rectangular first row.
  const first = rows[0]!;
  const values = rows.map((row) =>
    row.map((cell) => rendered.get(cell)!.trim().replaceAll("|", "\\|").replaceAll("\n", " ")),
  );
  if (!first.some((cell) => cell.tag === "th")) {
    values.unshift(first.map(() => ""));
  }
  values.splice(
    1,
    0,
    first.map(() => "---"),
  );
  return (
    "\n\n" +
    [...captions, values.map((row) => "| " + row.join(" | ") + " |").join("\n")].join("\n\n") +
    "\n\n"
  );
}

function serialize(
  node: Node,
  body: string,
  rendered: Map<Node, string>,
  base: string | null,
): string {
  const tag = node.tag;
  if (HEADINGS.has(tag)) {
    return body.trim() ? "\n\n" + "#".repeat(Number(tag[1])) + " " + body.trim() + "\n\n" : "";
  }
  const emphasis = EMPHASIS.get(tag);
  if (emphasis) {
    return wrapInline(body, emphasis);
  }
  if (["code", "kbd", "samp"].includes(tag)) {
    return inlineCode(node);
  }
  switch (tag) {
    case "pre":
      return fencedCode(node);
    case "a":
      return renderLink(node, body, base);
    case "img":
      return renderImage(node, base);
    case "ul":
    case "ol":
      return renderList(node, rendered);
    case "table":
      return renderTable(node, rendered);
    case "blockquote":
      return (
        "\n\n" +
        lines(body.trim())
          .map((line) => (line ? "> " + line : ">"))
          .join("\n") +
        "\n\n"
      );
    case "br":
      return "  \n";
    case "hr":
      return "\n\n---\n\n";
    case "dt":
      return "\n\n" + wrapInline(body.trim(), "**") + "\n";
    case "dd":
      return "\n" + body.trim() + "\n\n";
    default:
      return BLOCKS.has(tag) ? (body.trim() ? "\n\n" + body.trim() + "\n\n" : "") : body;
  }
}

function childParts(node: Node, rendered: Map<Node, string>): string[] {
  const parts: string[] = [];
  for (const [index, child] of node.children.entries()) {
    if (child instanceof Node) {
      parts.push(rendered.get(child)!);
      continue;
    }
    if (/^\p{White_Space}+$/u.test(child)) {
      const previous = node.children[index - 1];
      const following = node.children[index + 1];
      if (
        (previous instanceof Node && BLOCK_NEIGHBORS.has(previous.tag)) ||
        (following instanceof Node && BLOCK_NEIGHBORS.has(following.tag))
      ) {
        continue;
      }
    }
    parts.push(escape(child.replace(SPACE, " ")));
  }
  return parts;
}

export function render(selected: Node[], base: string | null): string {
  const output: string[] = [];
  for (const root of selected) {
    // Postorder fills the cache before a parent reads its children. Literal code
    // nodes read raw text directly and never consult descendant cache entries.
    const rendered = new Map<Node, string>();
    for (const node of postorder(root, LITERAL_TAGS)) {
      const parts = LITERAL_TAGS.has(node.tag) ? [] : childParts(node, rendered);
      rendered.set(node, serialize(node, joinParts(parts), rendered, base));
    }
    output.push(rendered.get(root)!);
  }
  const result = joinParts(output).trim();
  return result ? result + "\n" : "";
}

File: go/render.go
package htomd

import (
	"fmt"
	"math/big"
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"
)

var emphasis = map[string]string{"em": "*", "i": "*", "strong": "**", "b": "**", "s": "~~", "del": "~~", "strike": "~~"}
var literalTags = set("pre code kbd samp")
var blocks = set("p div article main section header footer figure figcaption address details summary dl")
var blockNeighbors = set("p div article main section header footer figure figcaption address details summary dl h1 h2 h3 h4 h5 h6 ul ol pre table blockquote hr li")
var blockMarker = regexp.MustCompile(`^(#{1,6}|[-+]|[0-9]+[.)]|[=~-]{3,})([\t\n\r\f ]|$)`)
var listStart = regexp.MustCompile(`^[+-]?[0-9]+$`)
var listValue = regexp.MustCompile(`^-?[0-9]{1,9}$`)

func escape(value string) string {
	value = strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;", "\\", "\\\\", "`", "\\`", "*", "\\*", "_", "\\_", "~", "\\~", "[", "\\[", "]", "\\]").Replace(value)
	lines := strings.Split(value, "\n")
	for i, line := range lines {
		trimmed := strings.TrimLeftFunc(line, isSpace)
		m := blockMarker.FindStringSubmatch(trimmed)
		if m == nil {
			continue
		}
		marker := m[1]
		if m[2] == "" && !(len(marker) >= 3 && strings.ContainsAny(marker, "=~-") && strings.Trim(marker, string(marker[0])) == "") {
			continue
		}
		escaped := "\\" + marker
		if marker[0] >= '0' && marker[0] <= '9' {
			escaped = marker[:len(marker)-1] + "\\" + marker[len(marker)-1:]
		}
		lines[i] = line[:len(line)-len(trimmed)] + escaped + trimmed[len(marker):]
	}
	return strings.Join(lines, "\n")
}
func longestRun(s string, ch rune) int {
	longest, current := 0, 0
	for _, c := range s {
		if c == ch {
			current++
			longest = max(longest, current)
		} else {
			current = 0
		}
	}
	return longest
}
func inlineCode(n *node) string {
	value := strings.NewReplacer("\r\n", " ", "\r", " ", "\n", " ").Replace(textContent(n, false))
	if value == "" {
		return ""
	}
	delimiter := strings.Repeat("`", longestRun(value, '`')+1)
	padding := ""
	if strings.HasPrefix(value, "`") || strings.HasSuffix(value, "`") || strings.HasPrefix(value, " ") && strings.HasSuffix(value, " ") && strings.TrimSpace(value) != "" {
		padding = " "
	}
	return delimiter + padding + value + padding + delimiter
}
func validLanguage(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if !unicode.IsLetter(c) && !unicode.IsNumber(c) && !strings.ContainsRune("_.+#-", c) {
			return false
		}
	}
	return true
}
func codeLanguage(n *node) string {
	candidates := append([]*node{n}, elements(n)...)
	if n.parent != nil {
		candidates = append(candidates, n.parent)
	}
	for _, c := range candidates {
		for _, token := range strings.Fields(c.attrs["class"]) {
			if strings.HasPrefix(token, "language-") || strings.HasPrefix(token, "lang-") || strings.HasPrefix(token, "highlight-") {
				value := strings.SplitN(token, "-", 2)[1]
				if validLanguage(value) {
					return value
				}
			}
		}
		if value := c.attrs["data-language"]; validLanguage(value) {
			return value
		}
	}
	return ""
}
func fencedCode(n *node) string {
	value := strings.NewReplacer("\r\n", "\n", "\r", "\n").Replace(textContent(n, false))
	fence := strings.Repeat("`", max(3, longestRun(value, '`')+1))
	ending := ""
	if !strings.HasSuffix(value, "\n") {
		ending = "\n"
	}
	return "\n\n" + fence + codeLanguage(n) + "\n" + value + ending + fence + "\n\n"
}
func joinParts(parts []string) string {
	result := []string{}
	for _, part := range parts {
		if part == "" {
			continue
		}
		if len(result) > 0 && strings.HasSuffix(result[len(result)-1], "\n") && strings.HasPrefix(part, "\n") {
			result[len(result)-1] = strings.TrimRight(result[len(result)-1], "\n")
			part = "\n\n" + strings.TrimLeft(part, "\n")
		}
		result = append(result, part)
	}
	return strings.Join(result, "")
}
func wrapInline(body, marker string) string {
	content := strings.TrimSpace(body)
	if content == "" {
		return body
	}
	leading, trailing := "", ""
	first, _ := utf8.DecodeRuneInString(body)
	last, _ := utf8.DecodeLastRuneInString(body)
	if isSpace(first) {
		leading = " "
	}
	if isSpace(last) {
		trailing = " "
	}
	return leading + marker + content + marker + trailing
}
func renderList(n *node, rendered map[*node]string) string {
	number := big.NewInt(1)
	start := strings.TrimSpace(n.attrs["start"])
	if listStart.MatchString(start) {
		number.SetString(start, 10)
	}
	items := []string{}
	for _, child := range elements(n) {
		if child.tag != "li" {
			continue
		}
		if n.tag == "ol" && listValue.MatchString(child.attrs["value"]) {
			number.SetString(child.attrs["value"], 10)
		}
		marker := "- "
		if n.tag == "ol" {
			marker = number.String() + ". "
		}
		body := strings.TrimSpace(rendered[child])
		if body != "" {
			lines := strings.Split(body, "\n")
			for i := 1; i < len(lines); i++ {
				if lines[i] != "" {
					lines[i] = strings.Repeat(" ", len(marker)) + lines[i]
				}
			}
			lines[0] = marker + lines[0]
			items = append(items, strings.Join(lines, "\n"))
		}
		number.Add(number, big.NewInt(1))
	}
	if len(items) == 0 {
		return ""
	}
	return "\n" + strings.Join(items, "\n") + "\n"
}
func tableRows(n *node) []*node {
	rows := []*node{}
	children := elements(n)
	stack := []*node{}
	for i := len(children) - 1; i >= 0; i-- {
		stack = append(stack, children[i])
	}
	for len(stack) > 0 {
		c := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if c.tag == "tr" {
			rows = append(rows, c)
		} else if c.tag == "thead" || c.tag == "tbody" || c.tag == "tfoot" {
			children = elements(c)
			for i := len(children) - 1; i >= 0; i-- {
				stack = append(stack, children[i])
			}
		}
	}
	return rows
}
func renderTable(n *node, rendered map[*node]string) string {
	rows := [][]*node{}
	for _, row := range tableRows(n) {
		cells := []*node{}
		for _, c := range elements(row) {
			if c.tag == "th" || c.tag == "td" {
				cells = append(cells, c)
			}
		}
		rows = append(rows, cells)
	}
	captions := []string{}
	for _, c := range elements(n) {
		if c.tag == "caption" {
			captions = append(captions, strings.TrimSpace(rendered[c]))
		}
	}
	simple := len(rows) > 0 && len(rows[0]) > 0
	for _, row := range rows {
		if len(row) != len(rows[0]) {
			simple = false
		}
		for _, c := range row {
			for _, a := range []string{"colspan", "rowspan"} {
				if value, ok := c.attrs[a]; ok && value != "1" {
					simple = false
				}
			}
			for _, child := range elements(c) {
				if set("table pre ul ol p")[child.tag] {
					simple = false
				}
			}
		}
	}
	lines := []string{}
	if !simple {
		for i, row := range rows {
			values := []string{}
			for _, c := range row {
				values = append(values, strings.ReplaceAll(strings.TrimSpace(rendered[c]), "\n", " "))
			}
			lines = append(lines, fmt.Sprintf("%d. %s", i+1, strings.Join(values, " — ")))
		}
	} else {
		values := [][]string{}
		header := false
		for _, c := range rows[0] {
			header = header || c.tag == "th"
		}
		if !header {
			values = append(values, make([]string, len(rows[0])))
		}
		for _, row := range rows {
			cells := []string{}
			for _, c := range row {
				cells = append(cells, strings.NewReplacer("|", "\\|", "\n", " ").Replace(strings.TrimSpace(rendered[c])))
			}
			values = append(values, cells)
		}
		separator := make([]string, len(rows[0]))
		for i := range separator {
			separator[i] = "---"
		}
		values = append(values[:1], append([][]string{separator}, values[1:]...)...)
		for _, row := range values {
			lines = append(lines, "| "+strings.Join(row, " | ")+" |")
		}
	}
	return "\n\n" + strings.Join(append(captions, strings.Join(lines, "\n")), "\n\n") + "\n\n"
}
func serialize(n *node, body string, rendered map[*node]string, base *string) string {
	tag := n.tag
	trimmed := strings.TrimSpace(body)
	if headings[tag] {
		if trimmed == "" {
			return ""
		}
		return "\n\n" + strings.Repeat("#", int(tag[1]-'0')) + " " + trimmed + "\n\n"
	}
	if marker, ok := emphasis[tag]; ok {
		return wrapInline(body, marker)
	}
	switch tag {
	case "code", "kbd", "samp":
		return inlineCode(n)
	case "pre":
		return fencedCode(n)
	case "a":
		href := safeURL(n.attrs["href"], base)
		if href == nil || *href == "" || trimmed == "" {
			return body
		}
		return "[" + trimmed + "](" + destination(*href) + ")"
	case "img":
		alt := escape(strings.TrimSpace(normalize(n.attrs["alt"])))
		source := safeURL(n.attrs["src"], base)
		if source == nil || *source == "" || n.attrs["src"] == "" {
			return alt
		}
		return "![" + alt + "](" + destination(*source) + ")"
	case "ul", "ol":
		return renderList(n, rendered)
	case "table":
		return renderTable(n, rendered)
	case "blockquote":
		lines := []string{}
		if trimmed != "" {
			for _, line := range strings.Split(trimmed, "\n") {
				if line == "" {
					lines = append(lines, ">")
				} else {
					lines = append(lines, "> "+line)
				}
			}
		}
		return "\n\n" + strings.Join(lines, "\n") + "\n\n"
	case "br":
		return "  \n"
	case "hr":
		return "\n\n---\n\n"
	case "dt":
		return "\n\n" + wrapInline(trimmed, "**") + "\n"
	case "dd":
		return "\n" + trimmed + "\n\n"
	}
	if blocks[tag] {
		if trimmed == "" {
			return ""
		}
		return "\n\n" + trimmed + "\n\n"
	}
	return body
}
func render(selected []*node, base *string) string {
	output := []string{}
	for _, root := range selected {
		// Visit each node once; literal serializers read their original descendants.
		order := []*node{}
		stack := []*node{root}
		for len(stack) > 0 {
			n := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			order = append(order, n)
			if !literalTags[n.tag] {
				stack = append(stack, elements(n)...)
			}
		}
		rendered := map[*node]string{}
		for i := len(order) - 1; i >= 0; i-- {
			n := order[i]
			parts := []string{}
			if !literalTags[n.tag] {
				for j, c := range n.children {
					if c.tag != "" {
						parts = append(parts, rendered[c])
						continue
					}
					if strings.TrimSpace(c.text) == "" && c.text != "" {
						if j > 0 && blockNeighbors[n.children[j-1].tag] || j+1 < len(n.children) && blockNeighbors[n.children[j+1].tag] {
							continue
						}
					}
					parts = append(parts, escape(normalize(c.text)))
				}
			}
			rendered[n] = serialize(n, joinParts(parts), rendered, base)
		}
		output = append(output, rendered[root])
	}
	result := strings.TrimSpace(joinParts(output))
	if result != "" {
		result += "\n"
	}
	return result
}
